### PR TITLE
Arm support improvement (part 1)

### PR DESCRIPTION
**Pull request by Snektron**
This PR implements the following functionality for better supporting ARM:
- Arm-eabi constants (std/os/bits/linux/arm-eabi.zig).
- Arm syscall convention (std/os/linux/arm-eabi.zig).
- Export `__aeabi_read_tp`, a function to get the thread pointer which LLVM expects to exist, if not linking with libc (std/os/linux/arm-eabi.zig). 
- Add TLS support for arm platforms (v6+) with a thread-pointer register. Targets without such a register can also make use of TLS, but require a different implementation of `__aeabi_read_tp`. A check is also added to see if a device supports such a register.

Generic 32-bit improvements:
- Fixed an overflow when calculating atime, mtime and ctime on 32-bit systems (std/fs/file.zig).
- Fixed a type error in std/debug.zig where an u64 representing the size of the executable would implicitly cast to an usize. I've added a simple `@intCast`. Though 32-bit platforms support files larger than 4 GB, im not sure if those can be mmap'ed all at once.
- Use MMAP2, getuid32-family and stat64-family syscalls when available.

Generic:
- When a debug-executable is sent to a device without source and panics, the debug panic handler would panic due to it not being able to find the source. It now simply doesn't print a source line if the source file cannot be found (std/debug.zig)
- Fix faulty arguments of preadv, preadv2, pwritev and pwritev2.

Several problems are still left open or were uncovered:
- Some types in libc are platform dependent, most notably off_t. In Musl it always 64-bit, even on 32-bit platforms. All functions support 64-bit file operations regardless of platform. In Glibc however, off_t is defined to be the CPU's native word size, unless a define to force it to 64-bit is given. See #3153.
- C abi stuff is not implemented yet. Even though a simple test program compiles, several standard library features are still missing for the test to compile. This depends on aforementioned problem, so i think that needs to be solved first.

Related: #2833

### CODE REVIEW

**Comment by Snektron** [comment]
(Note that some tests can't complete because of something in the existing master branch, as i just merged upstream/master into my branch. I'm not sure how to handle that best)

File: std/os/linux/tls.zig [hunk 1]
```diff
@@ -130,17 +133,29 @@ pub fn initTLS() void {
...
         }
     }
 
+    // If the cpu is arm-based, check if it supports the TLS register
+    if (builtin.arch == builtin.Arch.arm) {
+        if (at_hwcap & std.os.linux.HWCAP_TLS == 0) {
+            // If the CPU does not support TLS via a coprocessor register,
+            // a kernel helper function can be used instead on certain linux kernels.
+            // See linux/arch/arm/include/asm/tls.h and musl/src/thread/arm/__set_thread_area.c.
+            @panic("TODO: Implement ARM fallback TLS functionality");
```
Comment (andrewrk) [inline std/os/linux/tls.zig]: If there is no TLS program header, this panic would be irrelevant, right? Let's not panic unnecessarily if there is no thread local storage to set up.

File: std/debug.zig [hunk 2]
```diff
@@ -1053,7 +1053,7 @@ fn openSelfDebugInfoPosix(allocator: *mem.Allocator) !DwarfInfo {
     S.self_exe_file = try fs.openSelfExe();
     errdefer S.self_exe_file.close();
 
-    const self_exe_mmap_len = mem.alignForward(try S.self_exe_file.getEndPos(), mem.page_size);
+    const self_exe_mmap_len = mem.alignForward(@intCast(usize, try S.self_exe_file.getEndPos()), mem.page_size);
```
Comment (andrewrk) [inline std/debug.zig]: I think this should be `std.math.cast` and then if it doesn't fit then return either `error.InvalidDebugInfo` or a new error code that means we can't support such a large file. As it stands, this invokes undefined behavior in release-fast & release-small builds for 32 bit targets if the executable file is larger than 4 GiB.

Comment (andrewrk) [reply std/debug.zig]: cc @marler8997

File: std/os/linux/arm-eabi.zig [hunk 3]
```diff
@@ -0,0 +1,89 @@
+pub fn syscall0(number: usize) usize {
+    return asm volatile ("svc #0"
+        : [ret] "={r0}" (-> usize)
+        : [number] "{r7}" (number)
+    );
```
[the remark this answers is not in the document]
Comment (Snektron) [reply std/os/linux/arm-eabi.zig]: Yes, i see it's missing on AArch64 as well.

File: std/os/linux/arm-eabi.zig [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+/// This matches the libc clone function.
+pub extern fn clone(func: extern fn (arg: usize) u8, stack: usize, flags: u32, arg: usize, ptid: *i32, tls: usize, ctid: *i32) usize;
+
+// LLVM calls this when the read-tp-hard feature is set to false. Currently, there is no way to pass
+// that to llvm via zig, see https://github.com/ziglang/zig/issues/2883.
+// LLVM expects libc to provide this function as __aeabi_read_tp, so it is exported if needed from special/c.zig.
+pub nakedcc fn getThreadPointer() usize {
```
[the remark this answers is not in the document]
Comment (Snektron) [reply std/os/linux/arm-eabi.zig]: I think it should rather be a normal function, like the syscall ones. That should probably do, i'm not sure why i added the `nakedcc`.

**Comment by Snektron** [comment]
That should be every requested change implemented.

File: std/os/linux/tls.zig (hunk 1 above)
Comment (andrewrk) [reply std/os/linux/tls.zig]: I mean, can this check go below line 173 below?

```zig
     if (tls_phdr) |phdr| {
```

File: std/os/linux/tls.zig [hunk 5]
```diff
@@ -118,6 +118,9 @@ pub fn setThreadPointer(addr: usize) void {
                 : [addr] "r" (addr)
             );
         },
+        .arm => |arm| {
+            _ = std.os.linux.syscall1(std.os.linux.SYS_set_tls, addr);
```
Comment (daurnimator) [inline std/os/linux/tls.zig]: Should this check for an error and `@panic`?

Comment (Snektron) [reply std/os/linux/tls.zig]: Not as far as i'm aware. I've found no documentation for the syscall, and in the linux source it always returns 0:
https://github.com/torvalds/linux/blob/5ad18b2e60b75c7297a998dea702451d33a052ed/arch/arm/kernel/traps.c#L638-L640

Comment (daurnimator) [reply std/os/linux/tls.zig]: All syscalls can return errors via e.g. `seccomp` shenanigans.

**Pull request merged**